pkg/vault/contracts/authorizer/TimelockAuthorizerMigrator.sol

```diff
@@ -0,0 +1,166 @@
+        GRANT_PERMISSION_ACTION_ID = keccak256(abi.encodePacked(id, TimelockAuthorizer.grantPermissions.selector));
+        REVOKE_PERMISSION_ACTION_ID = keccak256(abi.encodePacked(id, TimelockAuthorizer.revokePermissions.selector));
+    }
+
+    /**
+     * @dev Migrate roles from the old authorizer to the new one
+     * @param n Number of roles to migrate, use 0 to migrate all the remaining ones
+     */
+    function migrate(uint256 n) external {
+        _beforeMigrate();
+        _migrate(n == 0 ? rolesData.length : n);
+        _afterMigrate();
+    }
+
+    function _migrate(uint256 n) internal {
+        uint256 i = migratedRoles;
+        uint256 to = Math.min(i + n, rolesData.length);
+        for (; i < to; i++) _migrate(rolesData[i]);
+        migratedRoles = i;
+    }
+
+    function _migrate(OldRoleData memory roleData) internal {
+        // In the old authorizer, each role has an admin role which might have been granted to many admins.
+        // In order to replicate that in the new authorizer, we need to grant permission for
+        // `TimelockAuthorizer.grantPermissions` to the same list of admins on each target.
+        // First, the contract iterates over the accounts that had the role granted in the old authorizer, granting
+        // the permission for the same role for the specified target in the new authorizer.
+        _migrate(roleData.role, roleData.target);
+
+        // Finally, the contract iterates over the list of role admins to grant `TimelockAuthorizer.grantPermissions`
+        // over the specified target. Note that this step may overlap with other roles data since this permission
+        // is at a general level for each target and not at a per-action level. However, the new authorizer ignores
+        // the request if the permission was granted already.
+        bytes32 adminRole = oldAuthorizer.getRoleAdmin(roleData.role);
+        _migrate(adminRole, roleData.target);
+    }
+
+    function _migrate(bytes32 role, address target) internal {
+        address[] memory wheres = _arr(target);
+        bytes32[] memory actionIds = _arr(role);
+        uint256 membersCount = oldAuthorizer.getRoleMemberCount(role);
+        for (uint256 i = 0; i < membersCount; i++) {
+            address member = oldAuthorizer.getRoleMember(role, i);
+            newAuthorizer.grantPermissions(actionIds, member, wheres);
+        }
+    }
+
+    function _beforeMigrate() internal view {
+        // Execute only once before the migration starts
+        if (migratedRoles > 0) return;
+
+        // Ensure the migrator contract has authority to change the vault's authorizer
+        bytes32 setAuthorizerId = IAuthentication(address(vault)).getActionId(IVault.setAuthorizer.selector);
+        bool canSetAuthorizer = oldAuthorizer.canPerform(setAuthorizerId, address(this), address(vault));
+        require(canSetAuthorizer, "MIGRATOR_CANNOT_SET_AUTHORIZER");
+    }
+
+    function _afterMigrate() internal {
```

Comment:
I'd be great if we had a simple getter that returned true or false depending on whether the migration has been completed or not.